### PR TITLE
Added some Type type predicates internally

### LINKED ISSUE #50005: Code cleanup: isIntersectionType & similar internally
# Suggestion

## 🔍 Search Terms

isIntersectionType type predicate internal source code

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Spinning https://github.com/microsoft/TypeScript/pull/49727#discussion_r910199321 into its own issue: there are a lot of type assertions internal to TypeScript that seem to predate type predicates.

It'd be nice to test out this this kind of change to the internal TypeScript source code:

```diff
- if (type.flags & TypeFlags.Intersection) {
+ if (isIntersectionType(type)) {
-     for (const subType of (type as IntersectionType).types) {
+     for (const subType of type.types) {
```

## 📃 Motivating Example

n/a

## 💻 Use Cases

See changes in https://github.com/microsoft/TypeScript/pull/49727.

Edit: ...and see #50010 for a draft PR.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Adds some type predicates with names like `is*Type` and uses them in a bunch of them internally. Note that while the diff shows slightly increased number of lines, the total number of _characters_ in checker.ts+utilities.ts goes from 3198453 to 3190463 (-7990).

Fixes #50005.

Throwing this up as a draft for reference. I'd be interested to see the results of a perf run...

### CODE REVIEW

**Comment by jakebailey** [comment]
When you get this to build, I'll happily do a perf run; I've been wanting to introduce these helpers for a while too.

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Ah, I should have waited for tests to pass before running perf, oops.

**Comment by JoshuaKGoldberg** [comment]
Yup, sorry for the false starts @jakebailey - it's ready now! 😄

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11900,23 +11900,23 @@ namespace ts {
...
+                    resolveUnionTypeMembers(type);
                 }
-                else if (type.flags & TypeFlags.Intersection) {
-                    resolveIntersectionTypeMembers(type as IntersectionType);
+                else if (isIntersectionType(type)) {
+                    resolveIntersectionTypeMembers(type);
                 }
                 else {
-                    Debug.fail("Unhandled type " + Debug.formatTypeFlags(type.flags));
+                    Debug.fail("Unhandled type " + Debug.formatTypeFlags((type as unknown as Type).flags));
```
Comment (jakebailey) [inline src/compiler/checker.ts:11910]: Why is this cast required? What is `type` here? `never`?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5326,7 +5326,7 @@ namespace ts {
                     const typeId = type.id;
                     const isConstructorObject = getObjectFlags(type) & ObjectFlags.Anonymous && type.symbol && type.symbol.flags & SymbolFlags.Class;
                     const id = getObjectFlags(type) & ObjectFlags.Reference && (type as TypeReference).node ? "N" + getNodeId((type as TypeReference).node!) :
```
Comment (jakebailey) [inline src/compiler/checker.ts:5328]: No `isTypeReference`? Or did that cause problems? (Or, is this PR just not exhaustive?)

**Review by jakebailey — COMMENTED** [review]
~~@typescript-bot perf test this~~

Using a review comment was probably too clever and the bot didn't like it.

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Or, it just won't run.

File: src/compiler/checker.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply src/compiler/checker.ts:11910]: Yeah, it's `never`. I did a quick search for the error message in the issue tracker and couldn't find anything.

File: src/compiler/checker.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply src/compiler/checker.ts:5328]: It's just non-exhaustive.

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -7674,8 +7674,68 @@ namespace ts {
...
+
+    export function isTypeVariableType(type: Type): type is TypeVariable {
+        return !!(type.flags & TypeFlags.TypeVariable);
+    }
+
+    export function isUnionType(type: Type): type is UnionType {
+        return !!(type.flags & TypeFlags.Union);
+    }
+
+    export function isUnionOrIntersectionType(type: Type): type is IntersectionType | UnionType {
```
Comment (JoshuaKGoldberg) [inline src/compiler/utilities.ts:7737]: I also just found the `isUnionOrIntersection` member of `TypeObject`. Is that preferred, or...?

Comment (jakebailey) [reply src/compiler/utilities.ts:7737]: Is that the one that's only available in `services`?

Comment (JoshuaKGoldberg) [reply src/compiler/utilities.ts:7737]: Yes, that. I'm wondering if, if these exist, those member predicates might not be needed anymore?

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Ouch; if these are to be believed, it's a good 3% perf loss.

Wish we had node 16/18 too, in case v8 has improved.

**Comment by JoshuaKGoldberg** [comment]
Yeah 😕 that is unfortunate. I suppose I'll close this PR & issue as not worth it.

It would be nifty if someone were to get the Closure Compiler running on the TypeScript codebase. I wonder if single-line functions like these type predicates could be inlined...

**Comment by DanielRosenwasser** [comment]
@JoshuaKGoldberg do you want to try to reopen this and move all the predicate functions to the checker? I wonder if the cost is from the object indirection of each check thanks to `ts.isYaddaYadda`.

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Hm, not quite as bad anymore, but I'm having trouble distinguishing it from noise too.

**Comment by jakebailey** [comment]
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Seems like it's consistently on the order of 1-1.5% worse? Better than 3% regression of the previous version, at least.

**Comment by sandersn** [comment]
@jakebailey  sooooo..do we take this? I hoped that slimming checker.ts would make it run faster, but apparently the overhead of (some non-inlined) function calls is bigger.

**Comment by jakebailey** [comment]
I'm not sure. I of course like the safety and would have totally sent this PR myself, but it seems like a 1-2% loss is not so great (but, it's an improvement over the previous iteration). But, I also can't tell if it's actually a regression or if it's in the range of variance.

I was hoping that after the module transform, we could place these wherever, but it turns out that it wouldn't work because our emit for imports doesn't actually create a local binding, so we have to have it here.

**Comment by DanielRosenwasser** [comment]
Unfortunately I think 1-2% is not worth the convenience. Let's not risk it.

**Comment by JoshuaKGoldberg** [comment]
Idea: what about a transformation API specific to the TypeScript source that replaces `is*Type` functions with their `.type ===` checks inline...

**Comment by jakebailey** [comment]
On my module-ified TypeScript branch, I've been testing something like that, but haven't gotten it to work quite yet.

**Pull request closed without merging**